File: implementations/node-engram/frontmatter-parser.js
#!/usr/bin/env node
/** Development parser for the language-neutral Engram front-matter corpus. */
const fs = require('fs');

class ParseError extends Error {
  constructor(code, message) { super(message); this.code = code; }
}

function stripComment(value) {
  let quote = null;
  for (let i = 0; i < value.length; i++) {
    const ch = value[i];
    if (quote === '"' && ch === '\\') { i++; continue; }
    if (quote && ch === quote) { quote = null; continue; }
    if (!quote && (ch === '"' || ch === "'")) { quote = ch; continue; }
    if (!quote && ch === '#' && (i === 0 || /\s/.test(value[i - 1]))) return value.slice(0, i).trimEnd();
  }
  return value.trimEnd();
}

function findColon(value) {
  let quote = null;
  for (let i = 0; i < value.length; i++) {
    const ch = value[i];
    if (quote === '"' && ch === '\\') { i++; continue; }
    if (quote && ch === quote) { quote = null; continue; }
    if (!quote && (ch === '"' || ch === "'")) { quote = ch; continue; }
    if (!quote && ch === ':' && (i + 1 === value.length || /\s/.test(value[i + 1]))) return i;
  }
  return -1;
}

function forbiddenFeature(front) {
  for (const original of front.split(/\r?\n/)) {
    const trimmed = original.trim();
    if (trimmed.startsWith('%') || trimmed === '...' || /^<<\s*:/.test(trimmed)) return true;
    let quote = null;
    for (let i = 0; i < original.length; i++) {
      const ch = original[i];
      if (quote === '"' && ch === '\\') { i++; continue; }
      if (quote && ch === quote) { quote = null; continue; }
      if (!quote && (ch === '"' || ch === "'")) { quote = ch; continue; }
      if (!quote && ch === '#') break;
      if (!quote && '[]{}'.includes(ch)) return true;
      if (!quote && '&*!'.includes(ch) && (i === 0 || /[\s:\-]/.test(original[i - 1]))) return true;
    }
  }
  return false;
}

function scalar(token) {
  const value = token.trim();
  if (value === '') return null;
  if (value.startsWith('"')) {
    if (!value.endsWith('"')) throw new ParseError('yaml-invalid', 'unterminated double-quoted scalar');
    try { return JSON.parse(value); } catch (error) { throw new ParseError('yaml-invalid', error.message); }
  }
  if (value.startsWith("'")) {
    if (!value.endsWith("'")) throw new ParseError('yaml-invalid', 'unterminated single-quoted scalar');
    return value.slice(1, -1).replace(/''/g, "'");
  }
  if (value === 'null') return null;
  if (value === 'true') return true;
  if (value === 'false') return false;
  if (/^-?(?:0|[1-9][0-9]*)$/.test(value)) return Number(value);
  if (/^-?(?:0|[1-9][0-9]*)(?:\.[0-9]+(?:[eE][+-]?[0-9]+)?|[eE][+-]?[0-9]+)$/.test(value)) {
    const number = Number(value);
    if (!Number.isFinite(number)) throw new ParseError('yaml-invalid', 'number must be finite');
    return number;
  }
  return value;
}

function parseKey(token) {
  const value = scalar(token);
  if (typeof value !== 'string') throw new ParseError('yaml-non-string-key', 'front-matter mapping keys must be strings');
  if (value === '<<') throw new ParseError('yaml-forbidden-feature', 'YAML merge keys are not permitted');
  return value;
}

function parseBlock(lines, start, indent) {
  if (start >= lines.length || lines[start].indent !== indent) throw new ParseError('yaml-invalid', 'invalid indentation');
  return lines[start].content === '-' || lines[start].content.startsWith('- ')
    ? parseSequence(lines, start, indent)
    : parseMapping(lines, start, indent);
}

function parseMapping(lines, start, indent) {
  const result = {};
  let index = start;
  while (index < lines.length && lines[index].indent === indent && !(lines[index].content === '-' || lines[index].content.startsWith('- '))) {
    const entry = lines[index].content;
    const colon = findColon(entry);
    if (colon < 0) throw new ParseError('yaml-invalid', `mapping entry lacks colon: ${entry}`);
    const key = parseKey(entry.slice(0, colon).trim());
    if (Object.prototype.hasOwnProperty.call(result, key)) throw new ParseError('yaml-duplicate-key', `duplicate YAML key ${JSON.stringify(key)}`);
    const remainder = entry.slice(colon + 1).trim();
    index++;
    if (remainder !== '') {
      result[key] = scalar(remainder);
    } else if (index < lines.length && lines[index].indent > indent) {
      const parsed = parseBlock(lines, index, lines[index].indent);
      result[key] = parsed.value; index = parsed.next;
    } else {
      result[key] = '';
    }
    if (index < lines.length && lines[index].indent > indent) throw new ParseError('yaml-invalid', 'unexpected indentation');
  }
  return {value: result, next: index};
}

function parseSequence(lines, start, indent) {
  const result = [];
  let index = start;
  while (index < lines.length && lines[index].indent === indent && (lines[index].content === '-' || lines[index].content.startsWith('- '))) {
    const remainder = lines[index].content.slice(1).trim();
    index++;
    if (remainder === '') {
      if (index < lines.length && lines[index].indent > indent) {
        const parsed = parseBlock(lines, index, lines[index].indent);
        result.push(parsed.value); index = parsed.next;
      } else result.push(null);
    } else if (findColon(remainder) >= 0) {
      const childIndent = indent + 2;
      let end = index;
      while (end < lines.length && lines[end].indent > indent) end++;
      const synthetic = [{indent: childIndent, content: remainder}, ...lines.slice(index, end)];
      const parsed = parseMapping(synthetic, 0, childIndent);
      if (parsed.next !== synthetic.length) throw new ParseError('yaml-invalid', 'invalid sequence mapping indentation');
      result.push(parsed.value); index = end;
    } else {
      result.push(scalar(remainder));
      if (index < lines.length && lines[index].indent > indent) throw new ParseError('yaml-invalid', 'unexpected indentation after sequence scalar');
    }
  }
  return {value: result, next: index};
}

function parseRecord(file, maxRecordBytes) {
  const raw = fs.readFileSync(file);
  if (raw.length > maxRecordBytes) throw new ParseError('resource-limit', 'record exceeds max_record_bytes');
  if (raw.length >= 3 && raw[0] === 0xef && raw[1] === 0xbb && raw[2] === 0xbf) throw new ParseError('bom-not-allowed', 'UTF-8 byte-order mark is not permitted');
  let text;
  try { text = new TextDecoder('utf-8', {fatal: true}).decode(raw); }
  catch (error) { throw new ParseError('encoding-invalid', error.message); }
  if (/\r(?!\n)/.test(text)) throw new ParseError('line-ending-invalid', 'bare CR line ending is not permitted');
  const physical = text.split(/\r?\n/);
  if (physical[0] !== '---') throw new ParseError('delimiter-invalid', 'opening delimiter must be the first exact line');
  const closing = physical.indexOf('---', 1);
  if (closing < 0) throw new ParseError('delimiter-invalid', 'missing exact closing delimiter');
  const front = physical.slice(1, closing).join('\n');
  if (forbiddenFeature(front)) throw new ParseError('yaml-forbidden-feature', 'forbidden YAML feature');
  const lines = [];
  for (const original of front.split('\n')) {
    if (/^ *\t/.test(original)) throw new ParseError('yaml-invalid', 'tabs cannot be used for indentation');
    const content = stripComment(original.slice(original.match(/^ */)[0].length));
    if (content.trim() === '') continue;
    lines.push({indent: original.match(/^ */)[0].length, content});
  }
  if (lines.length === 0) throw new ParseError('front-matter-not-mapping', 'front matter must be a mapping');
  const parsed = parseBlock(lines, 0, lines[0].indent);
  if (parsed.next !== lines.length) throw new ParseError('yaml-invalid', 'unconsumed YAML content');
  if (Array.isArray(parsed.value) || parsed.value === null || typeof parsed.value !== 'object') throw new ParseError('front-matter-not-mapping', 'front matter must be a mapping');
  return parsed.value;
}

function main() {
  if (process.argv.length !== 4 || process.argv[2] !== 'parse') {
    process.stderr.write('usage: frontmatter-parser.js parse REQUEST.json\n'); process.exit(2);
  }
  const request = JSON.parse(fs.readFileSync(process.argv[3], 'utf8'));
  const base = {protocol_version: '1.0', case_id: request.case_id};
  let result;
  try {
    result = {...base, outcome: 'accepted', front_matter: parseRecord(request.record, Number(request.max_record_bytes || 1048576))};
  } catch (error) {
    if (!(error instanceof ParseError)) throw error;
    result = {...base, outcome: 'rejected', diagnostic: {code: error.code, message: error.message}};
  }
  process.stdout.write(JSON.stringify(result) + '\n');
}

main();

File: implementations/python-engram/frontmatter_parser.py
#!/usr/bin/env python3
"""Development parser for the language-neutral Engram front-matter corpus."""
from __future__ import annotations

import json
import re
import sys
from pathlib import Path
from typing import Any

import yaml
from yaml import tokens as yaml_tokens


class ParseError(Exception):
    def __init__(self, code: str, message: str):
        super().__init__(message)
        self.code = code


class EngramLoader(yaml.SafeLoader):
    pass


EngramLoader.yaml_implicit_resolvers = {}
EngramLoader.add_implicit_resolver("tag:yaml.org,2002:null", re.compile(r"^null$"), ["n"])
EngramLoader.add_implicit_resolver("tag:yaml.org,2002:bool", re.compile(r"^(?:true|false)$"), ["t", "f"])
EngramLoader.add_implicit_resolver("tag:yaml.org,2002:int", re.compile(r"^-?(?:0|[1-9][0-9]*)$"), list("-0123456789"))
EngramLoader.add_implicit_resolver(
    "tag:yaml.org,2002:float",
    re.compile(r"^-?(?:0|[1-9][0-9]*)(?:\.[0-9]+(?:[eE][+-]?[0-9]+)?|[eE][+-]?[0-9]+)$"),
    list("-0123456789"),
)


def unique_mapping(loader: EngramLoader, node: yaml.MappingNode, deep: bool = False) -> dict[str, Any]:
    result: dict[str, Any] = {}
    for key_node, value_node in node.value:
        key = loader.construct_object(key_node, deep=deep)
        if not isinstance(key, str):
            raise ParseError("yaml-non-string-key", "front-matter mapping keys must be strings")
        if key == "<<":
            raise ParseError("yaml-forbidden-feature", "YAML merge keys are not permitted")
        if key in result:
            raise ParseError("yaml-duplicate-key", f"duplicate YAML key {key!r}")
        result[key] = loader.construct_object(value_node, deep=deep)
    return result


EngramLoader.add_constructor("tag:yaml.org,2002:map", unique_mapping)


def parse_record(path: Path, max_record_bytes: int) -> dict[str, Any]:
    raw = path.read_bytes()
    if len(raw) > max_record_bytes:
        raise ParseError("resource-limit", "record exceeds max_record_bytes")
    if raw.startswith(b"\xef\xbb\xbf"):
        raise ParseError("bom-not-allowed", "UTF-8 byte-order mark is not permitted")
    try:
        text = raw.decode("utf-8")
    except UnicodeDecodeError as exc:
        raise ParseError("encoding-invalid", str(exc)) from exc
    if re.search(r"\r(?!\n)", text):
        raise ParseError("line-ending-invalid", "bare CR line ending is not permitted")
    lines = text.splitlines(keepends=True)
    if not lines or lines[0] not in ("---\n", "---\r\n"):
        raise ParseError("delimiter-invalid", "opening delimiter must be the first exact line")
    closing = next((i for i, line in enumerate(lines[1:], 1) if line.removesuffix("\n").removesuffix("\r") == "---"), None)
    if closing is None:
        raise ParseError("delimiter-invalid", "missing exact closing delimiter")
    front_matter = "".join(lines[1:closing])
    try:
        for token in yaml.scan(front_matter, Loader=EngramLoader):
            if isinstance(token, (yaml_tokens.TagToken, yaml_tokens.AnchorToken, yaml_tokens.AliasToken)):
                raise ParseError("yaml-forbidden-feature", "tags, anchors, and aliases are not permitted")
            if isinstance(token, (yaml_tokens.DirectiveToken, yaml_tokens.DocumentStartToken, yaml_tokens.DocumentEndToken)):
                raise ParseError("yaml-forbidden-feature", "directives and document markers are not permitted")
            if isinstance(token, (yaml_tokens.FlowMappingStartToken, yaml_tokens.FlowSequenceStartToken)):
                raise ParseError("yaml-forbidden-feature", "flow collections are not permitted")
        value = yaml.load(front_matter, Loader=EngramLoader)
    except ParseError:
        raise
    except yaml.YAMLError as exc:
        raise ParseError("yaml-invalid", str(exc)) from exc
    if not isinstance(value, dict):
        raise ParseError("front-matter-not-mapping", "front matter must be a mapping")
    return value


def main() -> int:
    if len(sys.argv) != 3 or sys.argv[1] != "parse":
        print("usage: frontmatter_parser.py parse REQUEST.json", file=sys.stderr)
        return 2
    request = json.loads(Path(sys.argv[2]).read_text(encoding="utf-8"))
    base = {"protocol_version": "1.0", "case_id": request["case_id"]}
    try:
        value = parse_record(Path(request["record"]), int(request.get("max_record_bytes", 1048576)))
        result = {**base, "outcome": "accepted", "front_matter": value}
    except ParseError as exc:
        result = {**base, "outcome": "rejected", "diagnostic": {"code": exc.code, "message": str(exc)}}
    print(json.dumps(result, ensure_ascii=False, separators=(",", ":")))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
